libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Conway/NecessaryAndSufficient.hs

```diff
@@ -0,0 +1,373 @@
+assertWith n Remove x = assertExplain ["Status Remove " ++ show n ++ " (" ++ show x ++ ")"] TruePred
+
+testRule ::
+  ( STS (EraRule s a)
+  , BaseM (EraRule s a) ~ ShelleyBase
+  , HasSpec ConwayFn (Environment (EraRule s a))
+  , HasSpec ConwayFn (State (EraRule s a))
+  , HasSpec ConwayFn (Signal (EraRule s a))
+  ) =>
+  WitRule s a ->
+  Status ->
+  Specification ConwayFn (Environment (EraRule s a), State (EraRule s a), Signal (EraRule s a)) ->
+  (Signal (EraRule s a) -> String) ->
+  Gen Property
+testRule witrule status thespec nameF = do
+  triple@(env, state, signal) <- genFromSpec @ConwayFn thespec
+  pure $
+    monitorSpec
+      thespec
+      triple
+      ( goSTS
+          witrule
+          env
+          state
+          signal
+          ( \x ->
+              case (x, status) of
+                (Left _fails, Remove) -> classify True (nameF signal ++ " Fails") $ property True
+                (Left fails, Apply) -> counterexample ("Apply " ++ show signal ++ "\n" ++ show fails) (property False)
+                (Left _fails, Invert _) -> property False -- The monitor(\ _eval -> expectFailure) in assertWith, should catch this.
+                (Right _newdstate, Remove) -> classify True (nameF signal ++ " We got lucky, it randomly succeeds.") $ property True
+                (Right _newdstate, _) -> classify True (nameF signal) $ property True
+          )
+      )
+
+go1, go2, go3 :: Status -> IO ()
+go1 status = quickCheck (testRule (DELEG Conway) status (delegCertSpec status initPParams) delegName)
+go2 status = quickCheck (testRule (POOL Conway) status (poolRuleSpec status initPParams) poolName)
+go3 status = quickCheck (testRule (GOVCERT Conway) status (govRuleSpec @ConwayFn status) govName)
+
+delegProp, poolProp, govProp :: Status -> Gen Property
+delegProp status = testRule (DELEG Conway) status (delegCertSpec status initPParams) delegName
+poolProp status = testRule (POOL Conway) status (poolRuleSpec status initPParams) poolName
+govProp status = testRule (GOVCERT Conway) status (govRuleSpec @ConwayFn status) govName
+
+neccessaryAndSufficientTests :: Spec
+neccessaryAndSufficientTests =
+  describe "NeccessaryAndSufficient" $ do
+    prop "DELEG apply" $ delegProp Apply
+    prop "DELEG remove" $ delegProp Remove
+    prop "DELEG Invert 1" $ delegProp (Invert 1)
+    prop "DELEG Invert 2" $ delegProp (Invert 2)
+    prop "DELEG Invert 3" $ delegProp (Invert 3)
+    prop "DELEG Invert 4" $ delegProp (Invert 4)
+
+    prop "POOL apply" $ poolProp Apply
+    prop "POOL remove" $ poolProp Remove
+    prop "POOL Invert 1" $ poolProp (Invert 1)
+    prop "POOL Invert 2" $ poolProp (Invert 2)
+    prop "POOL Invert 3" $ poolProp (Invert 3)
+    prop "POOL Invert 4" $ poolProp (Invert 4)
+
+    prop "GOVCERT apply" $ govProp Apply
+    prop "GOVCERT remove" $ govProp Remove
+    prop "GOVCERT Invert 1" $ govProp (Invert 1)
+    prop "GOVCERT Invert 2" $ govProp (Invert 2)
+    prop "GOVCERT Invert 3" $ govProp (Invert 3)
```

Comment:
I think having to number each predicate that we want to flip *and* having to make a separate test for each number is a bit too much. Could this numbering be done behind the scenes somehow so that whoever is writing these constraints doesn't need to keep track of that?